### ISSUE TITLE
Parallelization of replica exchange simulations

### ISSUE THREAD

**Issue by cwalker7** [shirtsgroup/cg_openmm#12]
The replica exchange simulations in the example (done with Yank 0.21.2) seem to be running only in serial. Only 1 process is running when I ran a 24 replica example from a slurm batch script on Research Computing requesting 24 tasks on 1 node. The Yank code can be run with mpi (http://getyank.org/latest/running.html#parallelization) - it would be nice to have many or all replicas running in parallel.

**Comment by mrshirts** [comment]
Yeah, we definitely want to get the replicas running in parallel.  

Yank 0.21.2 is from 4/20/2018.  We should be using more up-to-date version.  
Starting from Yank 0.25 (Nov 8, 2019), the replica exchange functionality was moved to opemmtools, so I would suggest switching to openmmtools 0.18.1 for this functionality https://openmmtools.readthedocs.io/en/0.18.1/).  This would also remove Yank requirements, I believe (simplifying dependencies).

Once the code is up to date, then we can ask on the forums what settings need to be adjusted to get parallel replicas running correctly.

**Comment by cwalker7** [comment]
Yeah I agree we should update to openmmtools 0.18.1 and then worry about the parallelization.

**Comment by cwalker7** [comment]
I tried the advice from Andrea about setting the context cache, on my udpdate_depencies branch with openmmtools 0.19.0 and all packages at the latest versions. The commands that worked were a little different:

`from simtk import openmm
from openmmtools.cache import global_context_cache
global_context_cache.platform = openmm.openmm.Platform.getPlatformByName('CPU')`

But when I ran the replica_exchange.py example using mpirun I get this error repeated for each processor related to not finding the topology:

`Traceback (most recent call last):
  File "replica_exchange.py", line 124, in <module>
    positions=positions,
  File "/projects/chwa2399/software/anaconda/envs/cg_openmm_5_29/lib/python3.6/site-packages/foldamers-1.12-py3.6.egg/foldamers/cg_model/cgmodel.py", line 388, in __init__
    self.topology = build_topology(self, use_pdbfile=True)
  File "/projects/chwa2399/software/anaconda/envs/cg_openmm_5_29/lib/python3.6/site-packages/cg_openmm-1.0-py3.6.egg/cg_openmm/build/cg_build.py", line 370, in build_topology
FileNotFoundError: [Errno 2] No such file or directory: 'topology_source.pdb'
`

I get this whether or not I manually set the context cache - it's launching on the right number of processors, so I think the problem is with the cg_openmm code.

**Comment by mrshirts** [comment]
Each process was reading and writing a file ("topology_souorce.pdb") that had the same name, then deleting it.  I wrote to a temporary file instead (each process gets it's own), and that fixed it.  Maybe not the best solution, but works. I can successfully parallelize up to 12 processes on my laptop (which has 8 cores), and it runs to completion.

There were some other issues (simulation processing not parallelizable), so i need to clean up some other things before I file a PR, but I can run up to 12 processes in parallel now fine.

**Comment by mrshirts** [comment]
OK, PR #16 should fix this.  One should be able to mpriun -np [num processors] ./run_replica_exchage.py and parallelize all the replicas.  Would be interesting to see the scaling for a long run.

**Comment by cwalker7** [comment]
I tested out the new replica exchange code and was able to run 24 replicas with 24 tasks using mpirun (on CURC), and create the pdb files afterwards. The simulations finish but there is definitely something wrong with the bonds and replica exchange energies - some bonds are continuously increasing (going from 7.5 A to as high as ~70 A after 100ps simulation!) as if there is no restoring force. Could also be an issue with reporting the right coordinates for each replica.

**Comment by mrshirts** [comment]
Looks like the "some bonds are continuously increasing" problem fixed in #24
